lib/Parse/ParseExpr.cpp

```diff
@@ -2905,9 +2910,67 @@ ExprResult Parser::ParseBoundsExpression() {
     SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
 
   PT.consumeClose();
+
+  if (StartsRelativeBoundsClause(Tok) && !Result.isInvalid()) {
+    if (ParseRelativeBoundsClause(Result.get())) {
+      Diag(Tok, diag::err_unexpected_return_relative_bounds_clause);
+      Result = ExprError();
+    }
+  }
+
   return Result;
 }
 
+bool Parser::ParseRelativeBoundsClause(Expr * Expr) {
+  bool IsError = false;
+  RelativeBoundsClause *RelativeClause = nullptr;
+  ExprResult ConstExpr;
+
+  IdentifierInfo *Ident = Tok.getIdentifierInfo();
+  SourceLocation BoundsKWLoc = Tok.getLocation();
+  ConsumeToken();
+
+  // Skip until right paren and do nothing if it is not range bounds expr.
+  // for example bounds(none, none + 5). Currently, it is NullaryBoundsExpr
+  if (!(cast<BoundsExpr>(Expr)->isRange())) {
+    SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
+    return IsError;
+  }
+
+  BalancedDelimiterTracker PT(*this, tok::l_paren);
+
+  if (PT.expectAndConsume(diag::err_expected_lparen_after,
+                          Ident->getNameStart())) {
+    SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
+    IsError = true;
+    return IsError;
+  }
+
+  if (Ident == Ident_rel_align) {
+    TypeResult ResultTy = ParseTypeName();
+    if (ResultTy.isInvalid()) {
+      SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
+      IsError = true;
+    } else
+      RelativeClause = Actions.ActOnRelativeTypeBoundsClause(
+          BoundsKWLoc, ResultTy.get(), Tok.getLocation());
+  } else {
+    ConstExpr = Actions.VerifyIntegerConstantExpression(
+        ParseConstantExpression().get());
```

Comment:
ParseConstantExpression could return an invalid result.   You need code of the form
```
ConstExpr = ParseConstantExpression();
if (!ConstExpr.isInvalid()) {
  ConstExpr = Actions.VerifyIntegerConstantExpression(ConstExpr.get());
}